File: Service/PostalcodeService.vb
Imports BO
Imports DAL
Imports Facade

Public Class PostalcodeService
    Implements IPostalcodeService
    Public Function GetPostalcodeById(id As Integer) As GetResponse(Of PostalCodeBO) Implements IPostalcodeService.GetPostalcodeById
        Dim response As New GetResponse(Of PostalCodeBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetPostalcodeDAO

        Dim _entity = dao.GetById(id)
        Dim postalcode As New PostalCodeBO

        Dim err = PostalcodeTranslator.TranslateEntityToBO(_entity, postalcode)
        If err = ErrorCode.Success Then
            response.Value = postalcode
        Else
            response.AddError(err.ToString())
        End If
        Return response
    End Function
    Public Function GetPostalcodeByCountry(CountryId As Integer) As GetResponse(Of PostalCodeBO) Implements IPostalcodeService.GetPostalcodeByCountry
        Dim response As New GetResponse(Of PostalCodeBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetPostalcodeDAO()
        Dim entities = dao.GetNoTracking()
        For Each _entity In entities
            Dim bo As New PostalCodeBO()
            bo.PostcodeId = _entity.PostcodeID
            bo.Postcode = _entity.Postcode

            response.AddValue(bo)
        Next
        Return response
    End Function
    Public Function GetPostalcodeByCountryAndSearchstring(CountryId As Integer, Searchstring As String) As GetResponse(Of PostalCodeBO) Implements IPostalcodeService.GetPostalcodeByCountryAndSearchstring
        Dim response As New GetResponse(Of PostalCodeBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetPostalcodeDAO()
        Dim query = PostalcodeQuery.GetCountryQuery(CountryId)
        Dim subquery = PostalcodeQuery.GetCityContainsQuery(Searchstring)
        subquery = subquery.Or(PostalcodeQuery.GetPostalCodeStartsWithQuery(Searchstring))
        query = query.And(subquery)
        Dim entities = dao.GetNoTracking().Where(query)
        For Each _entity In entities
            Dim bo As New PostalCodeBO()
            bo.PostcodeId = _entity.PostcodeID
            bo.Postcode = _entity.Postcode
            bo.Gemeente = _entity.Gemeente

            response.AddValue(bo)
        Next
        Return response
    End Function

    Public Function GetPostalcodeByCountryCodeAndPostalcode(Countrycode As String, Postalcode As String) As GetResponse(Of PostalCodeBO) Implements IPostalcodeService.GetPostalcodeByCountryCodeAndPostalcode
        'Dim response As New GetResponse(Of PostalCodeBO)
        'Dim uow As New UnitOfWork(False)
        'Dim dao = uow.GetPostalcodeDAO()
        'Dim entities = dao.GetNoTracking.Where(Function(m) m.Country.LandISOCode = Countrycode And m.Country
        'For Each _entity In entities
        '    Dim bo As New PostalCodeBO()
        '    bo.PostcodeId = _entity.PostcodeID
        '    bo.Postcode = _entity.Postcode
        '    bo.Gemeente = _entity.Gemeente

        '    response.AddValue(bo)
        'Next
        'Return response
    End Function
End Class

File: ServiceCore/PostalcodeService.cs
using BOCore;
using FacadeCore;
using DALCore;
using DALCore.Models;
using ServiceCore.Translators;
using DALCore.Query;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace ServiceCore
{
    public class PostalcodeService : IPostalcodeService
    {
        private readonly UnitOfWorkCore _uow;

        public PostalcodeService(UnitOfWorkCore uow)
        {
            _uow = uow;
        }

        public GetResponse<PostalCodeBO> GetPostalcodeById(int id)
        {
            var response = new GetResponse<PostalCodeBO>();

            var entity = _uow.PostalCodes.GetById(id);   // pas naam aan als jouw UoW-property anders heet
            if (entity == null)
            {
                response.AddError("postalcode not found");
                return response;
            }

            var bo = new PostalCodeBO();
            var err = PostalcodeTranslator.TranslateEntityToBO(entity, bo);
            if (err == ErrorCode.Success) response.Value = bo;
            else response.AddError(err.ToString());

            return response;
        }

        public GetResponse<PostalCodeBO> GetPostalcodeByCountry(int countryId)
        {
            var response = new GetResponse<PostalCodeBO>();

            var entities = _uow.PostalCodes
                .GetNoTracking()
                .Where(PostalCodeQuery.GetCountryQuery(countryId));

            foreach (var e in entities)
            {
                var bo = new PostalCodeBO
                {
                    PostcodeId = e.PostcodeId,
                    Postcode = e.Postcode,
                    Gemeente = e.Gemeente
                };
                response.AddValue(bo);
            }
            return response;
        }

        public async Task<GetResponse<PostalCodeBO>> GetPostalcodeByCountryAndSearchstring(int countryId, string searchstring)
        {
            var response = new GetResponse<PostalCodeBO>();

            var query = PostalCodeQuery.GetCountryQuery(countryId);
            var sub = PostalCodeQuery.GetCityContainsQuery(searchstring)
                       .Or(PostalCodeQuery.GetPostalCodeStartsWithQuery(searchstring));

            var list = await _uow.PostalCodes
                .GetNoTracking()
                .Where(query.And(sub))
                .Select(e => new PostalCodeBO
                {
                    PostcodeId = e.PostcodeId,
                    Postcode = e.Postcode,
                    Gemeente = e.Gemeente
                })
                .ToListAsync(); // <— materialiseren

            foreach (var bo in list)
            {
                response.AddValue(bo);
            }
            return response;
        }
    }
}
